### ISSUE TITLE
practical issue with xml attributes syntax

### ISSUE THREAD

**Issue by seb-odoo** [odoo/owl#373]
After playing a bit with more complex examples, I have a suggestion about the syntax to declare components in xml, which I feel important since it is at the heart of the framework.

Indeed it looks clean for the small examples in the doc, but as soon as there are a lot of props/events to pass, I feel it becomes problematic.

Here are the problems I have with attributes:
- conceptually, there is a mix up of attributes with different semantics
     - `class`/`style` which are passed to child
     - events
     - "the rest" passed as props (potential conflicting names, need to know all the reserved keywords, ...)
- it is harder to validate a document (DTD), since the set of possible attributes is dynamic/infinite
- it is harder to override/xpath, especially events to add modifiers: the previous attribute has to be replaced by a new one, hardcoding the new name based on the previous one
- strings passed as props are likely not translated in Odoo, requiring a prealable t-set to achieve that
- visually, it is extremely hard to properly write many attributes:
    - all on the same line: it becomes unreadable if there are many of them
    - splitting lines: messes up the indent, because the attributes are on the same indent as the child block, etc

My suggestion to all of those problems would be to use sub-elements instead of attributes to define props and events.

To define a prop: use `t-set` (or `t-prop` if we want a new keyword), similar to what is done for defining sub-variables when doing `t-call`.

For events use `t-on` + `t-method` and possible children `t-modifier`.

Here is an example
```xml
<AutocompleteInput
    class="o_MessagingMenu_mobileNewMessageInput"
    customClass="id + '_mobileNewMessageInputAutocomplete'"
    isFocusOnMount="true"
    placeholder="mobileNewMessageInputPlaceholder"
    select="_onMobileNewMessageInputSelect"
    source="_onMobileNewMessageInputSource"
    t-on-o-hide.stop="_onHideMobileNewMessage"/>
```
would become
```xml
<AutocompleteInput class="o_MessagingMenu_mobileNewMessageInput">
    <t t-on="o-hide" t-method="_onHideMobileNewMessage">
        <t t-modifier="stop"/>
    </t>
    <t t-set="customClass" t-value="id + '_mobileNewMessageInputAutocomplete'"/>
    <t t-set="isFocusOnMount" t-value="true"/>
    <t t-set="placeholder" t-value="mobileNewMessageInputPlaceholder"/>
    <t t-set="select" t-value="_onMobileNewMessageInputSelect"/>
    <t t-set="source" t-value="_onMobileNewMessageInputSource"/>
</AutocompleteInput>
```
It takes a bit more characters to write, but does not have any of the above mentioned problems:
- indent is not messed up
- it is clear what is an event, a modifier, a prop, or "the rest", without having to know magical names/format
- easy to xpath, easy to validate structure
- possible to `<t t-set="key">value</t>` to translate it

**Comment by seb-odoo** [comment]
To conclude my idea:
- attribute names should never be dynamic,
- and we should never have to write many attributes on the same element.

**Comment by ged-odoo** [comment]
- hmm, this looks incompatible with the slot system (https://github.com/odoo/owl/blob/master/doc/component.md#slots)
- Also, I disagree on the fact that attribute names should never be dynamic.  They should usually not be dynamic, but there are use cases for dynamic attributes (this is necessary for some generic components).  
- "all the reserved keywords": you make is sound like there are many! I think that there are only two: `class` and `style`, which you mentioned...

But the big issue is how to make this work with slots...

**Comment by seb-odoo** [comment]
When I look at the `slot` example, it looks very much like `content` and `footer` are glorified props, which could benefit from the suggested syntax too?

If I understood correctly, the only difference is that they are rendered on the context of the parent and not the child, because they are declared as `t-slot` instead of `t-raw`, but the rest looks the same?

**Comment by ged-odoo** [comment]
as we discussed, I think that this is not easy to do, does not bring enough value to Owl, and we are under pressure to stabilize Owl. If this is important enough for you, maybe we can discuss someday some possible changes/update for a later Owl version. In the meantime, I will close this issue.

**Comment by seb-odoo** [comment]
I feel this one is indeed important, but we can postpone the discussion, hopefully not until a point in time where such change becomes out of question for stability reasons.
